source-snowflake/capture.go

```diff
@@ -0,0 +1,650 @@
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"slices"
+	"strconv"
+	"strings"
+
+	cerrors "github.com/estuary/connectors/go/connector-errors"
+	boilerplate "github.com/estuary/connectors/source-boilerplate"
+	"github.com/estuary/connectors/sqlcapture"
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	"github.com/jmoiron/sqlx"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	initialCloneSequenceNumber = 0
+	metadataProperty           = "_meta"
+)
+
+// Emit a checkpoint after every N change documents, so that partial
+// progress can be made across connector restarts. This is a tuning
+// constant which is only a variable so it can be lowered for tests.
+var partialProgressCheckpointSpacing = 10000
+
+type snowflakeSourceMetadata struct {
+	sqlcapture.SourceCommon
+
+	SequenceNumber int `json:"seq" jsonschema:"description=The sequence number of the staging table from which this document was read"`
+	TableOffset    int `json:"off" jsonschema:"description=The offset within that staging table at which this document occurred"`
+}
+
+func (snowflakeDriver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) error {
+	log.Debug("connector started")
+
+	var checkpoint = new(captureState)
+	if len(open.StateJson) > 0 {
```

Comment:
Might need to check for a state value of `{}` here or otherwise account for that possibility to avoid ending up with a `nil` `checkpoint.Streams` if there is no previously persisted state.

Reply:
Good catch, I've changed the logic to be `if checkpoint.Streams == nil { checkpoint.Streams = make(...) }` which should ensure that it's always initialized correctly.